### DISCUSSION TITLE
Example usage with cmd2

### DISCUSSION THREAD

**Discussion by TaihouKai** [CrowdStrike/Falcon-Toolkit#19] [Q&A]
I understand how Falcon-Toolkit works with command replay script, but I have no idea how does it work with cmd2.
May I have an example of how it works with cmd2? 

My knowledge to cmd2 is quite basic, e.g.:
```
from cmd2 import Cmd

class App(Cmd):
    """Sample"""

if __name__ == '__main__':
    import sys
    c = App()
    sys.exit(c.cmdloop())
```

For example, how to connect to some host (e.g. `-f Hostname=123456`), and run some RTR command (e.g., `ls .`)?

**Comment by ChristopherHammond13** [comment]
Hey! Thank you for opening the discussion.

I am currently out on annual leave and all my example code is on my work laptop. I will get some examples polished up w/b 3 January and publish them here.

In the meantime, please check [this page](https://cmd2.readthedocs.io/en/latest/features/scripting.html#python-scripts) out from the Cmd2 docs. Note that the app name is `rtr`, so commands would be executed like this:

```
rtr('ls .')
```

There is no need to subclass Cmd2. The scripts can be much simpler :) This is, at least, if you want to use the `run_pyscript` command to execute complex scripts with decision making logic within the shell.

For a more simple script that doesn't require Python scripting, you can just make a file that look like this:

```
ls .
quit
```

Then, invoke with `falcon shell -s script_filename`, or use the `run_script` shell command.

Let me know if this helps?

**Reply by TaihouKai** [reply]
Thank you so much for your detailed answer!
I'm sorry that I think I didn't understand correctly:
I'm reading the cmd2 doc you provided, does it mean like I need to run the following script using run_pyscript in `(Cmd)` ...
`(Cmd)`
```
(Cmd) run_pyscript foo.py
```
`foo.py`
```
rtr('ls .')
```
...? Cmd2 says `name 'rtr' is not defined`. I guess I missed something.

Again, thank you very much!

---

BTW: after `get /path/to/file`, is there anyway to automatically download the file?

**Reply by ChristopherHammond13** [reply]
Hey @TaihouKai, thank you for your patience here! I have found a script I wrote that will do exactly what you have described here. Please let me know if this helps you?

```python
"""
Automated file download script
For use with RTR Tookit
Usage: run_pyscript AutoGet.py <file path> <output directory>

Author: Chris Hammond <Chris.Hammond@crowdstrike.com>
"""
import sys

from time import sleep


def main():
    get_file_path = sys.argv[1]
    output_dir = sys.argv[2]

    # Download the file specified
    rtr(f"get \"{get_file_path}\"")

    # Start the loop
    max_iterations = 10
    sleep_time = 30

    batch_get_successful_requests = self.last_batch_get_successful_requests

    print(f"Downloading {batch_get_successful_requests} successful gets")

    iteration = 1

    while (
        self.last_batch_get_completed_uploads < batch_get_successful_requests and
        iteration <= max_iterations
    ):
        print(f"Iteration {iteration} of {max_iterations}. Sleeping for {sleep_time}s")
        sleep(sleep_time)
        rtr("get_status")
        iteration += 1

    rtr(f"download -e \"{output_dir}\"")
    print("Done!")


if len(sys.argv) != 3:
    script_filename = sys.argv[0]
    print(f"Usage: run_pyscript {script_filename} file_path output_directory")
else:
    main()
```

**Reply by TaihouKai** [reply]
Thank you so much for your help!
I get the following error:
```
(Cmd) run_pyscript rtr.py foo foo
Traceback (most recent call last):
  File "<string>", line 47, in <module>
  File "<string>", line 18, in main
NameError: name 'rtr' is not defined
```

I think I still have some setup problems...

**Reply by ChristopherHammond13** [reply]
Hey @TaihouKai, it works on this end. Just note that you need to run the Python script from within an RTR shell launched via the `falcon` command, and not through Cmd2 directly. For example, you might want to do something like this

```shell
$ falcon shell -f Hostname=SYSTEM1
...
C:\> run_pyscript AutoGet.py C:\target\file\path.txt /Users/TaihouKai/Downloads
```

Let me know how this goes?

**Reply by TaihouKai** [reply]
It works. Thank you so much!
